File: array/10-squares-of-a-sorted-array/squares-of-a-sorted-array.py
from types import List


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        i, j, pos = 0, n - 1, n - 1
        while i <= j:
            if nums[i] ** 2 > nums[j] ** 2:
                res[pos] = nums[i] ** 2
                i += 1
            else:
                res[pos] = nums[j] ** 2
                j -= 1
            pos -= 1
        return res

File: array/10-squares-of-a-sorted-array/squares-of-a-sorted-array.go
package main

import "sort"

type Solution1 struct{}

func (s1 Solution1) sortedSquares(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	i, j, pos := 0, n-1, n-1
	for i <= j {
		if nums[i]*nums[i] > nums[j]*nums[j] {
			res[pos] = nums[i] * nums[i]
			i++
		} else {
			res[pos] = nums[j] * nums[j]
			j--
		}
		pos--
	}
	return res
}

type Solution2 struct{}

func (s2 Solution2) sortedSquares(nums []int) []int {
	for i := range nums {
		nums[i] = nums[i] * nums[i]
	}
	sort.Ints(nums)
	return nums
}
